drivers: clock_control: nrf driver reimplementation

nordic-krch opened the pull request:
Clock_control has been re-implemented to support new clock control API and calibration algorithm which optionally uses temperature sensor and skips calibration on no temperature change.

- Added new API implementation
- Added calibration algorithm (calibration support extracted to separate file)
- Added test suite
- Updated usage of clock _control API in nordic drivers, removed hacky usage of subsys parameter.

-----------
Fixes #12114

nordic-krch committed:
drivers: clock_control: nrf: reimplementation including API updates

Reimplementation of clock control driver for nrf platform. It includes
latest API changes: asynchronous starting and getting clock status.

Additionally, it implements calibration algorithm which optionally
skips calibration based on no temperature change. Internal temperature
sensor is used for that.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: clock_control: Add test suite

Added test suite for clock control driver. It covers latest API update
for starting clock asynchronously and getting status.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
bluetooth: controller: Align nrf link layer to new clock control API

Align to use clock_control_get_status for blocking clock start.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: Align nrf counter, timer and usb driver to new clock_control

Align drivers to use new clock control API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

aescolar commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -12,269 +12,308 @@
 
-static int m16src_start(struct device *dev, clock_control_subsys_t sub_system)
+/* Function returns current interrupt priority or 0xff if in thread mode */
+static int z_arch_irq_current_prio(void)
```
This function would seem quite out of place here. Shouldnt it be in arch/arm/core/irq_mange.c  or so?

nordic-krch replied:
yes, that was my forgotten intention. See #17301.

aescolar committed:
nrf52_bsim: Require v1.8 of the HW models

The incomming new NRF clock driver requires extra nrf HAL
and drivers functions, which are only supported in the HW models
after version 1.8

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,311 @@
+
+/* Clock instance structure */
+struct nrf_clock_control {
+	atomic_t ref;		/* Users counter */
```
Why is it defined as `atomic_t` when atomic API is not used for this field?

nordic-krch replied:
changed to `s32_t`

anangl reviewed (changes requested):
Just some initial comments. I'll need some more time to dive into `nrf_power_clock.c`.

aescolar commented:
@nordic-krch In https://github.com/aescolar/zephyr/tree/clock_control_calibration_fix  you have this branch with a couple of updates for simulation and rebased on top of #17301 . With it the nrf52_bsim works nicely

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,191 @@
+
+LOG_MODULE_DECLARE(clock_control, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#define TEMP_SENSOR_NAME COND_CODE_1(TEMP_NRF5, (CONFIG_TEMP_NRF5_NAME), (NULL))
```
fixed

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,191 @@
+static struct device *temp_sensor; /* Handler to temperature sensor device. */
+
+static void measure_temperature(struct k_work *work);
+struct k_work temp_measure_work = {
```
fixed

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,191 @@
+	sensor_sample_fetch(temp_sensor);
+	sensor_channel_get(temp_sensor, SENSOR_CHAN_DIE_TEMP, &sensor_val);
+
+	return sensor_value_to_float(&sensor_val);
```
changed to work in 0,25 degree units. Also modified kconfig option to use same units which allows to chose fractional temperature diff that will trigger calibration.

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,191 @@
+		} else {
+			cal_state = CAL_TEMP_REQ;
+			k_work_submit_to_queue(&k_sys_work_q,
+					       &temp_measure_work);
```
done

nordic-krch commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -85,7 +85,7 @@ int z_clock_driver_init(struct device *device)
 	}
 
-	clock_control_on(clock, (void *)CLOCK_CONTROL_NRF_K32SRC);
+	clock_control_on(clock, 0);
```
done

nordic-krch commented on ext/hal/nordic/drivers/nrf_radio_802154/platform/clock/nrf_802154_clock_zephyr.c:
```diff
@@ -63,8 +63,10 @@ void nrf_802154_clock_hfclk_start(void)
+    clock_control_on(clk_m16, NULL);
+    /* Block until clock is ready. */
+    while (clock_control_get_status(clk_m16, NULL) !=
+		CLOCK_CONTROL_STATUS_ON) {}
```
done

nordic-krch commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -27,29 +27,32 @@ struct temp_nrf5_data {
-	 * it should block.
-	 */
-	r = clock_control_on(data->hfclk_dev, (void *)1);
+	r = clock_control_async_on(data->hfclk_dev, 0, &clk_data);
```
done

nordic-krch commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -27,29 +27,32 @@ struct temp_nrf5_data {
 	struct device *hfclk_dev;
 };
 
+void hfclk_on_callback(struct device *dev, void *user_data)
```
done

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,311 @@
 #include <nrf_clock.h>
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+#include <logging/log.h>
+#if defined(CONFIG_USB_NRF52840) || NRF_POWER_HAS_POFCON
```
now nrf_power.h is always included.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,311 @@
+#define INT_MASK \
+	(NRF_CLOCK_INT_HF_STARTED_MASK | NRF_CLOCK_INT_LF_STARTED_MASK \
+	COND_CODE_1(CONFIG_USB_NRF52840, \
+		    (|NRF_POWER_INT_USBDETECTED_MASK | \
```
done

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
+struct nrf_clock_control_config {
+	u32_t start_tsk;	/* Clock start task */
+	u32_t started_evt;	/* Clock started event */
+	u32_t stop_tsk;		/* Clock stop task */
```
Why not use dedicated types for these fields, i.e. `nrf_clock_task_t` and `nrf_clock_event_t`? These are offsets, so you would also save some memory here (if the compiler is allowed to use short enums). Place the `stop_handler` first to avoid padding in this case.

nordic-krch replied:
changed.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
 
-	NRF_CLOCK->EVENTS_LFCLKSTARTED = 0;
+#if defined(CONFIG_USB_NRF52840) || NRF_POWER_HAS_POFCON
+static bool power_event_check_and_clean(u32_t evt, u32_t intmask)
```
By using `nrf_clock_event_t` for `evt` you could prevent such mistakes as the one with `nrf_clock_event_clear` below.

nordic-krch replied:
ack

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
 
-	irq_unlock(imask);
+	if (data->ref > 0) {
+		return data->started ?
```
That's true but on the other hand one can question that result of get status is always volatile if someone that did not requested clock is checking it's state. Main purpose for having that function is to block until clock is ready. In that case if one requested clock then it will never be in `OFF` state.

I wouldn't like to complicate this function only to satisfy use case which will rather never occur. Maybe we should clarify `clock_control.h` the use case. Maybe we should even replace this function with something like `clock_control_wfo()` (wait for on)?

anangl replied:
What about changing it to:
```
if (data->started) {
	return CLOCK_CONTROL_STATUS_ON;
}
if (data->ref > 0) {
	return CLOCK_CONTROL_STATUS_STARTING;
}

return CLOCK_CONTROL_STATUS_OFF;
```

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
-		nrf_clock_task_trigger(NRF_CLOCK_TASK_HFCLKSTART);
+	if (data->ref == 0) {
+		DBG(dev, "Stopping");
+		sys_slist_init(&data->list);
```
if clock was stopped before started then we need to clear the list. otherwise after starting again list will not be empty and unwanted callback may be called on next clock start.

anangl replied:
> if clock was stopped before started then we need to clear the list

Right. I haven't thought about this case. It's not that obvious, so perhaps a short comment here would be a good thing?

nordic-krch committed:
tests: drivers: clock_control: Add test for nRF clock calibration

Added test for LF clock calibration when RC source is used.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,311 @@
-		/* Set the Calibration Timer Initial Value */
-		NRF_CLOCK->CTIV = 16;	/* 4s in 0.25s units */
+	if (ret) {
+		nrf_clock_event_clear(evt);
```
done

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -286,197 +325,33 @@ static inline void power_event_cb(nrf_power_event_t event)
-		NRF_CLOCK->EVENTS_CTTO = 0;
+	if (clock_event_check_and_clean(NRF_CLOCK_EVENT_HFCLKSTARTED,
+					NRF_CLOCK_INT_HF_STARTED_MASK)) {
+		struct device *hfclk_dev = &DEVICE_NAME_GET(clock_nrf5_m16src);
```
done

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -286,197 +310,33 @@ static inline void power_event_cb(nrf_power_event_t event)
-		power_event_cb(NRF_POWER_EVENT_USBDETECTED);
+	if (clock_event_check_and_clean(NRF_CLOCK_EVENT_LFCLKSTARTED,
+					NRF_CLOCK_INT_LF_STARTED_MASK)) {
+		clkstarted_handle(&DEVICE_NAME_GET(clock_nrf5_k32src));
```
done

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
 #include <errno.h>
 #include <sys/atomic.h>
```
includes cleaned up.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
+	struct nrf_clock_control *data = dev->driver_data;
 
-	m16src_grd = 1U;
+#if 0
```
removed that temporarily

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,190 @@
+static struct device *temp_sensor; /* Handler to temperature sensor device. */
+
+static void measure_temperature(struct k_work *work);
+K_WORK_DEFINE(temp_measure_work, measure_temperature);
```
added

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
-	 * resource guard.
+static int clock_control_init(struct device *hfclk_dev)
+{
+	/* TODO: Initialization will be called twice, once for 32KHz and then
```
removed.

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,190 @@
+
+void nrf_clock_control_calibration_init(struct device *dev)
+{
+	/* errata 56 */
```
added errata description. in fact number here was wrong. fixed that.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
-
-	/* release resource guard */
-	m16src_grd = 0U;
+	nrf_clock_int_enable(INT_MASK);
```
done

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
+struct nrf_clock_control {
+	s32_t ref;		/* Users counter */
+	sys_slist_t list;	/* List of users requesting callback */
+	bool started;		/* When set clock is started */
```
rephrased.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
+
+	switch (get_status(dev, sub_system)) {
+	case CLOCK_CONTROL_STATUS_ON:
+		irq_unlock(key);
```
fixed.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
+		 * the second time.
+		 */
+		if (data) {
+			if (data->node.next) {
```
that was risky. added function for checking if element is in the list. not relying on that anymore.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -9,272 +9,296 @@
+	switch (get_status(dev, sub_system)) {
+	case CLOCK_CONTROL_STATUS_ON:
+		irq_unlock(key);
+		if (data && data->cb) {
```
added assert to check that. In general `data->cb` cannot be NULL, if user does not want callback then `data` is NULL. If `data != NULL` then `data->cb != NULL`, too.

nordic-krch committed:
tests: bluetooth: mesh: Fix RAM shortage failure for bbc_microbit

Test for bbc_microbit was failing due to RAM usage. Reduced main
stack size and turned off temperature algorithm to fit.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -1,280 +1,327 @@
+
+/* returns true if clock stopping or starting can be performed. If false then
+ * start/stop will be deferred and performed later on by handler owner.*/
+typedef bool (*nrf_clock_stop_handler_t)(struct device *dev);
```
The name of this type is no longer adequate, now that it is used also for checking if starting can be performed.

nordic-krch replied:
renamed to `nrf_clock_handler_t`

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -1,280 +1,327 @@
+	nrf_clock_task_t stop_tsk;	/* Clock stop task */
+};
+
+#define INT_MASK \
```
No longer used.

nordic-krch replied:
removed

anangl commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,271 @@
+
+	irq_unlock(key);
+
+	return false;
```
Is this really supposed to always return `false`?

nordic-krch replied:
no, fixed now.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -1,280 +1,327 @@
-	ARG_UNUSED(sub_system);
+			if (do_start) {
+				nrf_clock_event_clear(config->started_evt);
+				nrf_clock_task_trigger(config->start_tsk);
```
Where will this task be triggered in case `start_handler` returns `false`?

nordic-krch replied:
if start_handler returns false then it means that starting clock will be handled by module that provided handler. In that particular case start handler will returns false when it didn't yet stop the clock because calibration was ongoing. It will not call start because lfclk was not yet stopped (`CAL_ACTIVE_OFF` state).

anangl commented on include/drivers/clock_control/nrf_clock_control.h:
```diff
@@ -61,4 +59,46 @@
+void nrf_clock_control_calibration_isr(void);
+
+/**
+ * @brief Stop calibration.
```
Invalid description.

nordic-krch replied:
fixed

anangl commented on include/drivers/clock_control/nrf_clock_control.h:
```diff
@@ -61,4 +59,46 @@
+/**
+ * @brief Stop calibration.
+ *
+ * Function called when LFCLK RC clock is being started. It starts calibration
```
~~It starts calibration~~

anangl commented on include/drivers/clock_control/nrf_clock_control.h:
```diff
@@ -61,4 +59,46 @@
+ *
+ * @retval true if clock can be started.
+ * @retval false if clock was not stopped due to ongoing calibration and don't
+ *	   need to be started again because it is still on.
```
Currently this is not entirely true. The function returns `false` when `cal_state == CAL_ACTIVE_OFF` and this means that the calibration is ongoing but after it completes, the clock will be stopped.

nordic-krch replied:
it's actually true (at least now). If before calibration is done, calibration state will be changed to `idle` then in calibration done event clock will not be stopped.

That was initial intention but due to bug in the function it was misbehaving.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -1,280 +1,327 @@
 
-	nrf_clock_task_trigger(NRF_CLOCK_TASK_HFCLKSTOP);
+static int clock_control_init(struct device *hfclk_dev)
+{
```
`data->list` should be initialized here (and in `k32_dummy_init`).

nordic-krch replied:
done

anangl commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -0,0 +1,318 @@
+	*executed = true;
+}
+
+static void clock_on_callback2(struct device *dev, void *user_data)
```
No need for a second callback function when you are using different user data.

nordic-krch replied:
done

anangl commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -0,0 +1,318 @@
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	/* lock to prevent clock interrupt for fast starting clocks.*/
+	key = irq_lock();
```
This will only prevent handling of the interrupt. If a clock starts fast enough, it can still generate the interrupt before `clock_control_off` is called. In such case the interrupt will be handled (and the callback executed) right after `irq_unlock`.

nordic-krch replied:
that's why clearing started event after stopping the clock. Not sure how it will be done in other platforms but driver should ensure that interrupt will not be generated after stopping the clock before it got started.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -1,280 +1,327 @@
+				(NRF_POWER_INT_USBDETECTED_MASK |
+				NRF_POWER_INT_USBREMOVED_MASK |
+				NRF_POWER_INT_USBPWRRDY_MASK),
+				(0))));
```
Wouldn't such formatting:
```
	nrf_clock_int_enable(
		NRF_CLOCK_INT_HF_STARTED_MASK |
		NRF_CLOCK_INT_LF_STARTED_MASK |
		COND_CODE_1(CONFIG_USB_NRF52840,
			(NRF_POWER_INT_USBDETECTED_MASK |
			 NRF_POWER_INT_USBREMOVED_MASK |
			 NRF_POWER_INT_USBPWRRDY_MASK),
			(0)));
```
be a bit more readable?

nordic-krch replied:
done

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,258 @@
+};
+
+static enum nrf_cal_state cal_state; /* Calibration state. */
+static float prev_temperature; /* Previous temperature measurement. */
```
done

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,258 @@
+}
+
+/* Function reads from temperature sensor and converts to float. */
+static float get_temperature(void)
```
strange that there was no warning. fixed now.

nordic-krch commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -1,280 +1,327 @@
+	nrf_clock_stop_handler_t start_handler; /* Called before start */
+	nrf_clock_stop_handler_t stop_handler; /* Called before stop */
+	nrf_clock_task_t start_tsk;	/* Clock start task */
+	nrf_clock_event_t started_evt;	/* Clock started event */
```
there is this case when you are stopping the clock before it got started. Then during stopping clock may get started and started event is set. If it is not cleared after calling stop task then after leaving stop procedure interrupt will be generated as if clock started but it is already stopped.

One of the test case covers that so when i removed event clearing test was failing. Another way to workaround that would be to check actual register status in the interrupt. I think that clearing that event is simpler solution.

anangl replied:
I see. This clearing seems to be deserving a comment. It's not that obvious why the STARTED event is cleared right after the STOP task is triggered.

nordic-krch replied:
added comment.

nordic-krch commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,45 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+
+config CLOCK_CONTROL_NRF_CALIBRATION_PERIOD
+	int "Calibration opportunity period (in 250ms units)"
+	default 1
```
done

nordic-krch commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,45 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+
+config CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP
+	int "Maximum number of calibration skips"
+	default 4
```
done

nordic-krch commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -0,0 +1,318 @@
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_STARTING, status,
```
removed locking and relaxed check:
just after starting expecting `starting` or `on` status, after busy for waiting startup time expecting `on`.

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,272 @@
+	int key = irq_lock();
+
+	cal_state = CAL_IDLE;
+	if (cal_state != CAL_ACTIVE_OFF) {
```
my bad. fixed.

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,272 @@
+		on_cal_done();
+	}
+
+	if (/*NRF_CLOCK_INT_CTSTOPPED_MASK &&*/
```
fixed

anangl commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,273 @@
+	 */
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_DONE);
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_CTTO);
+	nrf_clock_event_clear(NRF_CLOCK_EVENT_CTSTOPPED);
```
This should be inside `if (NRF_CLOCK_EVENT_CTSTOPPED) {}`.
It's not that harmful right not, as it clears `HFCLKSTARTED` if `CTSTOPPED` is not present, but...

nordic-krch replied:
it's actually not needed. I've added it because i had some issues which turns out to be not related to that. forgot to remove it then. Removed now.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -1,280 +1,333 @@
 
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
-static inline void power_event_cb(nrf_power_event_t event)
+static void pof_isr(void)
```
Do we actually need this? The `POFWARN` interrupt is not active, so this function actually does nothing but a little delay.

nordic-krch replied:
right, when POF support will be added then it makes sense. removed.

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -286,197 +339,33 @@ static inline void power_event_cb(nrf_power_event_t event)
-				ctto = 1U;
-			}
-#endif
+		if (!data->started) {
```
I think I already asked about this, but my question somehow got lost.
Is is done because of the anomaly 201? It should be commented properly then. Now it looks like a redundant check.

nordic-krch replied:
added comment about anomaly 201.

nordic-krch commented:
@pizi-nordic @cvinayak @ioannisg anyone would find some time to do the review? @anangl did quite extensive one but we need another to get that in.

ioannisg commented:
> 
> 
> @pizi-nordic @cvinayak @ioannisg anyone would find some time to do the review? @anangl did quite extensive one but we need another to get that in.

I think we should all review this; it is a major driver, affecting all systems in nRF platforms. I am going to do a pass; for now, I am a bit concerned whether this increases RAM usage... I saw, here, a ram-usage fix in one of nRF51 tests, and I started thinking about it. We should are to re-work the driver, we need to ensure this does not bring further overhead for the resource-constrained nrf51 SOC

pizi-nordic commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,272 @@
+	CAL_HFCLK_REQ,	/* HFCLK XTAL requested. */
+	CAL_TEMP_REQ,	/* Temperature measurement requested. */
+	CAL_ACTIVE,	/* Ongoing calibration. */
+	CAL_ACTIVE_OFF /* Ongoing calibration, off requested. */
```
Please align comments.

nordic-krch replied:
done

pizi-nordic commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,272 @@
+static bool clock_event_check_and_clean(u32_t evt, u32_t intmask)
+{
+	bool ret = nrf_clock_event_check(evt) &&
+			nrf_clock_int_enable_check(intmask);
```
Why we care clearing event only if IRQ is enabled?

nordic-krch replied:
why should we clean it always?

pizi-nordic replied:
Why we are cleaning event only when IRQ is enabled for it?

pizi-nordic commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,272 @@
+	return (s16_t)(4 * val->val1 + val->val2 / 250000);
+}
+
+/* Function reads from temperature sensor and converts to float. */
```
I do not see float in this function.

nordic-krch replied:
fixed

pizi-nordic commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,272 @@
+
+/* Function determines if calibration should be performed based on temperature
+ * measurement. Function is called from system work queue context. It is
+ * reading temperature from TEMP sensor and compares with last measurement.
```
You cannot use last measurement! This way you might never run the calibration if the temp changes slowly!

nordic-krch replied:
we might improve algorithm in the future, this is initial one. If temperature changes slowly then calibration will be forced after exceeding `CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP` skips.

pizi-nordic replied:
So please set the limit for this variable CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP.
IMHO fixing the algorithm is not a big deal...

nordic-krch replied:
@pizi-nordic please take a look. the diff is now calculated based on measurement done during last calibration.

pizi-nordic commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,272 @@
+		if (diff >= CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_TEMP_DIFF) {
+			start_calibration();
+		} else {
+			LOG_DBG("Calibration skipped - stable temperature.");
```
IMHO it might be a good idea to print the temperature difference here.

nordic-krch replied:
done

pizi-nordic commented on tests/drivers/clock_control/clock_control_api/nrf_lfclk_rc.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_ZTEST=y
+CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC=y
```
No new line at the end of file.

nordic-krch replied:
done

pizi-nordic commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -0,0 +1,310 @@
+			(status == CLOCK_CONTROL_STATUS_ON),
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	k_busy_wait(startup_us*1.5);
```
Why you are using multiplier here? Aren't deadlines specified in the devices array?

nordic-krch replied:
I want to take into account potential unalignment or imprecision of k_busy_wait.

pizi-nordic replied:
k_busy_wait is pretty accurate. IMHO it would be better to pass the values from the table directly.

nordic-krch replied:
removed `1.5*` and trimmed startup times.

pizi-nordic commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -0,0 +1,310 @@
+	}
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_true(status != CLOCK_CONTROL_STATUS_OFF,
```
Could we specify here what we expect not what we not expect?

nordic-krch replied:
at this point clock might be `STARTING` or `ON`. It's hard to determine. If clock is starting fast then it may be on already.

pizi-nordic replied:
So we should have `status == STARTING || status == ON`. This will be easier to read.

nordic-krch replied:
done

pizi-nordic commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -0,0 +1,310 @@
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	for (int i = 0; i < users; i++) {
+		err = clock_control_off(dev, 0);
```
Could we also check what happen if there will be more off requests than on requests?

nordic-krch replied:
added one more attempt expecting `-EALREADY`

pizi-nordic commented on tests/drivers/clock_control/clock_control_api/src/test_clock_control.c:
```diff
@@ -0,0 +1,310 @@
+
+	zassert_true(executed1, "%s: Expected flag to be true", dev_name);
+	zassert_true(executed2, "%s: Expected flag to be true", dev_name);
+}
```
IMHO we should also check if the clocks are really on.

nordic-krch replied:
It's hard to check that in generic way. Callback is the indicator

pizi-nordic commented on tests/drivers/clock_control/nrf_clock_calibration/src/test_nrf_clock_calibration.c:
```diff
@@ -0,0 +1,181 @@
+
+	clock_control_on(lfclk_dev, NULL);
+
+	/* It may be required to restart CT. Allow that by 1ms without lock */
```
CT?

nordic-krch replied:
replaced CT with calibration timer,

pizi-nordic commented on tests/drivers/clock_control/nrf_clock_calibration/src/test_nrf_clock_calibration.c:
```diff
@@ -0,0 +1,181 @@
+		}
+	}
+
+	zassert_within(cnt, 250, 10, "Expected 250ms period");
```
Am I correct assuming that the first calibration will start 250ms after clock startup? If so, shouldn't we start calibration as early as possible in order to provide stable clock?

nordic-krch replied:
i've changed the behavior. first calibration is performed immediately after starting LF clock.

pizi-nordic commented on tests/drivers/clock_control/nrf_clock_calibration/src/test_nrf_clock_calibration.c:
```diff
@@ -0,0 +1,181 @@
+	irq_unlock(key);
+
+	while (clock_control_get_status(hfclk_dev, NULL)
+			!= CLOCK_CONTROL_STATUS_ON) {
```
In my opinion we always should specify what we waiting for.

nordic-krch commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,47 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+	  Calibration is skipped when temperature change did not exceeded
+	  configured threshold. If number of consecutive skips reaches
+	  configured value then calibration is performed unconditionally. Set to
+	  0 to perform unconditional calibration periodically.
```
done

nordic-krch commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,47 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+	  configured value then calibration is performed unconditionally. Set to
+	  0 to perform unconditional calibration periodically.
+
+config CLOCK_CONTROL_NRF_CALIBRATION_TEMP_DIFF
```
i've rephrased it, please take another look.

nordic-krch commented:
@ioannisg 
> I am a bit concerned whether this increases RAM usage...

It's only couple of bytes and that brought your attention only because there was a test that was using 99,5% or RAM. There are dozens of places where ram is wasted. It's increased mainly when LF RC is used because of async data structure required (12 bytes).

pizi-nordic commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,51 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+	help
+	  Periodically, calibration action is performed. Action includes
+	  temperature measurement followed by clock calibration. Calibration may
+	  be skipped if temperature change (compared to previous measurement)
```
... compared to temperature of previous calibration ...

pizi-nordic commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,51 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+	int "Temperature change triggering calibration (in 0.25 degree units)"
+	default 2
+	help
+	  if temperature change between two consecutive measurements equals or
```
... since last calibration ...

pizi-nordic reviewed (commented):
Please update description in order to describe current behaviour.

nordic-krch commented:
Rebased.

@pabigot it's no longer blocked by #17301. 
@pizi-nordic long time ago you were soo close to approve, then vacation time came. Can you take another look?

pabigot commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,279 @@
+		ret = false;
+	}
+
+	cal_state = CAL_IDLE;
```
I don't have any faith that this unconditional setting of `cal_state` correctly maintains the state machine.  It could, if the call site does the necessary checking for other values, but the call site is in another source file and obscured by a couple layers of indirection through so it's not obvious.

I really think these functions should be private (see other comment).

nordic-krch replied:
Function are somehow private now (see other comment)

pabigot commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,51 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+
+config CLOCK_CONTROL_NRF_CALIBRATION_PERIOD
+	int "Calibration opportunity period (in 250ms units)"
+	default 1
```
Why is the default 1?  Last I looked the recommended interval was 4 s to maintain accuracy required for Bluetooth.  (That was some time back, so maybe it does need to be done at 4 Hz now.)

nordic-krch replied:
shouldn't then CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP = 1? With 2 that would result in 12 seconds if no temp change.

nordic-krch replied:
applied defaults as suggested (but set CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP = 1 to achieve calibration every 8s when no temperature change).

pabigot commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,51 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
-	  initially start running and automatically switch to crystal when
-	  ready.
+	  Calibration is skipped when temperature change since last calibration
+	  did not exceeded configured threshold. If number of consecutive skips
```
"~did not exceeded~was less than"

Typo and edge case error (it isn't skipped when the change equaled the threshold).

nordic-krch replied:
fixed.

pabigot commented on drivers/clock_control/Kconfig.nrf:
```diff
@@ -47,14 +47,51 @@ config CLOCK_CONTROL_NRF_K32SRC_EXT_FULL_SWING
+	  and number of consecutive skips did not exceeded
+	  CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP.
+
+config CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP
```
Do you really need this feature if the default CTIV is 16 (4 s) instead of 1 (0.25 s)?  I guess it does allow a maximum interval of 255*31.25 s, and allows rapid recalibration of the temperature changes at rates like 1 Cel / s but I'm not convinced the complexity of a two-level timeout counter is warranted.

pabigot reviewed (commented):
A few comments.  The complexity of this is a bit more than I can comprehend right now.  If the calibration material was in the same source file as the code that invokes it, it'd be easier to be convinced that the state machine maintenance is correct.  Also some of my comments on CTIV are made in ignorance of what other Nordic infrastructure has done.  The two-level counter+skip approach seems far more complex than it needs to be, but if you're re-implementing a standard approach from NCS or SDK or nrfx then I guess it's OK.

As a future enhancement it would be nice to have the temperature monitor pulled out to a separate API that could be shared between RC calibration and ADC calibration.

ioannisg commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2016-2019 Nordic Semiconductor ASA
+ * Copyright (c) 2016 Vinayak Kariappa Chettimada
```
Reason, here, may be that these contributions were done before Nordic was officially involved in Zephyr

ioannisg replied:
I mean, the reason for individual copyright

joerchan replied:
@ioannisg This is a new file in this PR.

ioannisg replied:
@joerchan , it was a rather general answer ;) There are such files in the controller subsystem.

nordic-krch commented on tests/bluetooth/mesh/microbit.conf:
```diff
@@ -50,3 +50,5 @@ CONFIG_BT_MESH_LABEL_COUNT=0
 #CONFIG_BT_MESH_DEBUG_ADV=y
 #CONFIG_BT_MESH_DEBUG_ACCESS=y
+
+CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP=0
```
fixed commit message

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,279 @@
+	int key;
+
+	temperature = get_temperature();
+	diff = temperature - prev_temperature;
```
done

nordic-krch commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -319,9 +319,12 @@ int lll_clk_on_wait(void)
+
+	while (clock_control_get_status(lll.clk_hf, NULL) !=
+			CLOCK_CONTROL_STATUS_ON) {
 		DEBUG_RADIO_XTAL(1);
```
done

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,279 @@
+
+	if (clock_event_check_and_clean(NRF_CLOCK_EVENT_DONE,
+					NRF_CLOCK_INT_DONE_MASK)) {
+		LOG_DBG("Calibration done.");
```
done

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,279 @@
+}
+
+/* Start calibration assuming that HFCLK XTAL is on. */
+static void start_calibration(void)
```
Added `void z_nrf_clock_calibration_force_start(void);` function for `nrf_clock_control.h` (which is considered nordic specific but public). This call will trigger calibration immediately.

nordic-krch commented on drivers/clock_control/nrf_clock_calibration.c:
```diff
@@ -0,0 +1,279 @@
+		}
+	}
+
+	irq_unlock(key);
```
done

nordic-krch commented:
It's been out for 3 months. There are 3 approvals. @carlescufi @ioannisg can we merge this then unless someone plans to review it soon?

aescolar commented:
For the record, I have no issue with this PR. The only file changed I own is perfectly fine. But as this is a Nordic only change, I do not believe I should approve it. I presume BT and USB maintainers would have a similar feeling as the BT and USB code changes are minor and circumstantial.

ioannisg commented:
> It's been out for 3 months. There are 3 approvals. @carlescufi @ioannisg can we merge this then unless someone plans to review it soon?

I would appreciate if @pabigot could re-review once more (and approve if possible). I guess we can merge this afterwards ;)

pabigot reviewed (commented):
My comments appear to have been addressed.  I haven't tested this or done an in-depth functional review so don't feel comfortable "approving" it, but I don't see any reason it should be kept out.

martintv commented:
I haven't looked much at the code, this switch to zephyr is all new to me, so I am still confused. So so this is more a question than a suggestion.  After this PR we will still use the Calibration timer in the NRF_CLOCK peripheral right? how easy would it be to switch to using some other source for the periodic timer to drive the calibration state machine in zephyr world? we always wanted to do that in SoftDevice, but never did. The calibration timer in in NRF_CLOCK has some big awkward weaknesses, since some register can not be written withing the same 32khz clock cycle as some things happens. Another reason to stop using the calibration timer, is that we want to remove it from the chip, to save silicon, if sw could do without it.

nordic-krch commented:
@martintv i don't see any obstacles. In the next step we can replace calibration timer with k_timer.

nordic-krch commented:
@ioannisg see @pabigot comment. What then blocks as from merging it?

ioannisg commented:
> 
> 
> @ioannisg see @pabigot comment. What then blocks as from merging it?

Your description is still blocking this PR @nordic-krch. 
#17301 is not reviewed/merged. Please, advice

nordic-krch commented:
@ioannisg i've updated description as this is no longer true. I don't use functionality which #17301 attempted to add.

ioannisg commented:
> 
> 
> @ioannisg i've updated description as this is no longer true. I don't use functionality which #17301 attempted to add.

OK, great, now this PR is not blocked-by-definition ;)
Could you rebase and run CI (PR hasn't been rebased for ~10 days, AFAICS)

nordic-krch commented:
@ioannisg, done.

carlescufi reviewed (approved):
I tested the following:

- `central_hr` on nRF51 DK
- `peripheral` (complete with bonding and reconnection) on nRF51 DK
- `central_hr` on BBC micro:bit
- `peripheral_hr` on BBC micro:bit

joerchan commented:
There is by default a lot of info lines being printed. Can we change the settings so that that doesn't happen?
```
[00:10:10.496,673] <inf> clock_control: CLOCK_16M: callback
[00:10:10.496,704] <inf> clock_control: CLOCK_16M: node 0x0000000
```

carlescufi commented:
> There is by default a lot of info lines being printed. Can we change the settings so that that doesn't happen?
> 
> ```
> [00:10:10.496,673] <inf> clock_control: CLOCK_16M: callback
> [00:10:10.496,704] <inf> clock_control: CLOCK_16M: node 0x0000000
> ```
> 
> Tested connections using 2 nrf52840. and 500 PPM, works fine.

I saw that as well, I assumed this was my config but now that @joerchan also brings it up I think this is worth cleaning up. @nordic-krch could you please disable those by default?

nordic-krch commented:
@carlescufi @joerchan removed those log calls.

@carlescufi I've reduced irq locking time. Additionally, i did measurement to check longest locking time in clock driver vs other lockings (used peripheral_hr sample for that). Clock locking is up to 1.4us while there are many many more longer lockings (up to 2.5ms which seems very odd). In general locking in clock is one of the shortest.

carlescufi commented:
> @carlescufi @joerchan removed those log calls.
> 
> @carlescufi I've reduced irq locking time. Additionally, i did measurement to check longest locking time in clock driver vs other lockings (used peripheral_hr sample for that). Clock locking is up to 1.4us while there are many many more longer lockings (up to 2.5ms which seems very odd). In general locking in clock is one of the shortest.

Thank you @nordic-krch. I will merge this as soon as CI passes.
